### PR TITLE
[RIP-70-3]Extract adaptive lock mechanism

### LINKED ISSUE #8442: [RIP-70] Optimizing Lock Mechanisms
### Before Creating the Enhancement Request

- [X] I have confirmed that this should be classified as an enhancement rather than a bug/feature.

### Summary

As concurrent systems grow more complex internally in RockteMQ, deploying effective lock management strategies is key to preserving performance. Our goal is to optimize the locking mechanism and improve performance.

I have written the proposal and you can see it by the link below:
https://shimo.im/docs/473QM8m8ByCn6l3w/

We will implement the proposed changes by 3 phases.

- [x]  1.Optimize the locking logic for message delivery to commitLog
        2.Spin optimal degree K retreat locking mechanism is introduced
        3.Optimize the back pressure mechanism of the client

- [x]  1.Adaptive lock is implemented initially
        2.Introduce other locking mechanisms
        3.Optimize message receiving logic

- [x]  1.Improved adaptive locks
        2.Provides tools for adjusting the lock mechanism
        3.Design the necessary tests

### Motivation

null

### Describe the Solution You'd Like

null

### Describe Alternatives You've Considered

null

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by 3424672656**
### Which Issue(s) This PR Fixes

Fixes #8442 

### How Did You Test This Change?
Test document ：https://shimo.im/docs/ZzkLMQ4RwwUa87AQ/

### CODE REVIEW

**Review by fuyou001 — CHANGES_REQUESTED** [review]
CR need

File: store/src/main/java/org/apache/rocketmq/store/lock/AdaptiveBackOffLockImpl.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+                    needSwap = true;
+                }
+            } else if (lock.getNumberOfRetreat(slot) * base * 3 / 2 <= tps) {
+                lock.adapt(false);
+            }
+            lock.setNumberOfRetreat(slot, 0);
+        } else {
+            if (tps <= this.tpsSwapCriticalPoint * 4 / 5) {
+                needSwap = true;
+            }
```
Comment (GenerousMan) [inline store/src/main/java/org/apache/rocketmq/store/lock/AdaptiveBackOffLockImpl.java:114]: Too many magic numbers, it should be optimized.

File: store/src/main/java/org/apache/rocketmq/store/lock/CollisionRetreatLock.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+import org.apache.rocketmq.store.config.MessageStoreConfig;
+
+import java.time.LocalTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class CollisionRetreatLock implements AdaptiveBackOffLock {
```
Comment (GenerousMan) [inline store/src/main/java/org/apache/rocketmq/store/lock/CollisionRetreatLock.java:27]: What is RetreatLock? Is it a misspelling of reentrant lock?

File: store/src/main/java/org/apache/rocketmq/store/lock/CollisionRetreatLock.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    private int optimalDegree;
+
+    private int initOptimalDegree;
+
+    private int maxOptimalDegree;
+
+    private final List<AtomicInteger> numberOfRetreat;
+
+    public CollisionRetreatLock() {
+        this.initOptimalDegree = 1000;
```
Comment (GenerousMan) [inline store/src/main/java/org/apache/rocketmq/store/lock/CollisionRetreatLock.java:40]: You'd better make these numbers as constants, like OPTIMAL_DEGREE = 1000, MAX_OPTIMAL_DEGREE = 10000

Comment (3424672656) [reply store/src/main/java/org/apache/rocketmq/store/lock/CollisionRetreatLock.java]: done~

File: store/src/main/java/org/apache/rocketmq/store/CommitLog.java [hunk 4]
```diff
@@ -95,6 +97,8 @@ public class CommitLog implements Swappable {
 
     protected final PutMessageLock putMessageLock;
 
+    protected final AdaptiveBackOffSpinLock adaptiveBackOffSpinLock;
```
Comment (RongtongJin) [inline store/src/main/java/org/apache/rocketmq/store/CommitLog.java:100]: 不需要成为对象的变量，在方法中new即可

Comment (3424672656) [reply store/src/main/java/org/apache/rocketmq/store/CommitLog.java]: done~

File: common/src/main/java/org/apache/rocketmq/common/BrokerConfig.java [hunk 5]
```diff
@@ -427,6 +427,11 @@ public class BrokerConfig extends BrokerIdentity {
     // if false, will still rewrite ck after max times 17
     private boolean skipWhenCKRePutReachMaxTimes = false;
 
+    /**
+     * Use AdaptiveBackOffLock
+     **/
+    private boolean useABSLock = false;
```
Comment (RongtongJin) [inline common/src/main/java/org/apache/rocketmq/common/BrokerConfig.java:433]: It would be better to put this config to useABSLock

Comment (3424672656) [reply common/src/main/java/org/apache/rocketmq/common/BrokerConfig.java]: done~

File: store/src/main/java/org/apache/rocketmq/store/lock/AdaptiveBackOffSpinLock.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+ * limitations under the License.
+ */
+package org.apache.rocketmq.store.lock;
+
+import org.apache.rocketmq.store.PutMessageLock;
+import org.apache.rocketmq.store.config.MessageStoreConfig;
+
+public interface AdaptiveBackOffSpinLock extends PutMessageLock {
+
+    void lock();
```
Comment (fuyou001) [inline store/src/main/java/org/apache/rocketmq/store/lock/AdaptiveBackOffSpinLock.java:24]: The PutMessageLock class includes methods for locking and unlocking.

File: store/src/main/java/org/apache/rocketmq/store/CommitLog.java [hunk 7]
```diff
@@ -130,7 +132,11 @@ protected PutMessageThreadLocal initialValue() {
                 return new PutMessageThreadLocal(defaultMessageStore.getMessageStoreConfig());
             }
         };
-        this.putMessageLock = messageStore.getMessageStoreConfig().isUseReentrantLockWhenPutMessage() ? new PutMessageReentrantLock() : new PutMessageSpinLock();
+
+        AdaptiveBackOffSpinLock adaptiveBackOffSpinLock = new AdaptiveBackOffSpinLockImpl();
```
Comment (fuyou001) [inline store/src/main/java/org/apache/rocketmq/store/CommitLog.java:136]: PutMessageLock adaptiveBackOffSpinLock = new AdaptiveBackOffSpinLockImpl() may be better

File: store/src/main/java/org/apache/rocketmq/store/lock/AdaptiveBackOffSpinLockImpl.java [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+                lock.adapt(false);
+            }
+            lock.setNumberOfRetreat(slot, 0);
+        } else {
+            if (tps * threadNum <= this.swapCriticalPoint * SWAP_SPIN_LOCK_RATIO) {
+                needSwap = true;
+            }
+        }
+
+        if (needSwap) {
```
Comment (fuyou001) [inline store/src/main/java/org/apache/rocketmq/store/lock/AdaptiveBackOffSpinLockImpl.java:131]: Add some comments to explain the principles and the rationale behind these numbers.

File: store/src/main/java/org/apache/rocketmq/store/lock/AdaptiveBackOffSpinLockImpl.java [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+
+    private int swapCriticalPoint;
+
+    private AtomicInteger currentThreadNum = new AtomicInteger(0);
+
+    private AtomicBoolean isOpen = new AtomicBoolean(true);
+
+    public AdaptiveBackOffSpinLockImpl() {
+        this.locks = new HashMap<>();
+        this.locks.put("Reentrant", new BackOffReentrantLock());
```
Comment (fuyou001) [inline store/src/main/java/org/apache/rocketmq/store/lock/AdaptiveBackOffSpinLockImpl.java:55]: These strings use class constants.

File: store/src/main/java/org/apache/rocketmq/store/lock/AdaptiveBackOffSpinLockImpl.java (hunk 8 above)
Comment (3424672656) [reply store/src/main/java/org/apache/rocketmq/store/lock/AdaptiveBackOffSpinLockImpl.java:144]: done~

File: store/src/main/java/org/apache/rocketmq/store/CommitLog.java (hunk 7 above)
Comment (3424672656) [reply store/src/main/java/org/apache/rocketmq/store/CommitLog.java]: done~

File: store/src/main/java/org/apache/rocketmq/store/lock/AdaptiveBackOffSpinLock.java (hunk 6 above)
Comment (3424672656) [reply store/src/main/java/org/apache/rocketmq/store/lock/AdaptiveBackOffSpinLock.java]: done~

**Review by fuyou001 — DISMISSED** [review]

**Review by GenerousMan — APPROVED** [review]
LGTM

**Review by RongtongJin — APPROVED** [review]

**Review by fuyou001 — APPROVED** [review]

**Review by RongtongJin — APPROVED** [review]

**Pull request merged**